torch/csrc/jit/codegen/cuda/fusion_segmenter.h

```diff
@@ -0,0 +1,365 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/fusion.h>
+#include <torch/csrc/jit/codegen/cuda/ir_base_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler_registry.h>
+
+#include <deque>
+#include <list>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//  FusionSegmentFinder
+//    Responsible for going through DAG and proposing things we could try to
+//    fuse together, calls "canGenerateCode" on these proposed segments to see
+//    if they are valid and we can generate code for them.
+//  FusionSegment
+//    A group of exprs that are segmented together
+//  FusionSegmentConnections
+//    Holds vals and what they connect. In other words it's a val that is an
+//    output of a FusionSegment "from" and an input of FusionSegment "to".
+//    There's nothing preventing from a val being between segments twice.
+//    TODO: make sure there's nothing wrong with segmentation on nodes that
+//    have the same value input twice. i.e. (B = A*A)
+
+// Selecting segments to propose is based on the theorem 4.2 in the paper which
+// makes sure when segment the segmented graph will be a DAG (assumes Fusion is
+// already a DAG). The segmentation code relies on assumptions of DAG-ness
+// during segmentation, meaning proposed merging of groups must maintain the DAG
+// property of the graph.
+//
+// Julien Herrmann, Yusuf Özkaya, Bora Uçar, Kamer Kaya, Umit Catalyurek.
+// Multilevel Algorithms for Acyclic Partitioning of Directed Acyclic Graphs.
+// SIAM Journal on Scientific Computing, Society for Industrial and Applied
+// Mathematics, 2019, 41 (4), pp.A2117-A2145. ff10.1137/18M1176865ff.
+// ffhal02306566f
+
+class SegmentedGroup;
+class SegmentCandidateFinder;
+
+// A directed edge on DAG,
+// Wrapper for values, edges between segmented groups which are made up
+// of Exprs. Multiple edges can exist between segmented groups.
+class SegmentedEdge {
+ public:
+  SegmentedEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val)
+      : from_(from), to_(to), val_(val) {}
+  SegmentedGroup* from_;
+  SegmentedGroup* to_;
+  Val* val_;
+
+  void print();
+};
+
+std::ostream& operator<<(std::ostream& os, const SegmentedEdge* edge);
+
+// Groups together expressions which create a segmented group
+// Can be used to produce fusions
+class TORCH_CUDA_API SegmentedGroup {
+ public:
+  SegmentedGroup() = default;
+
+  SegmentedGroup(Expr* expr) {
+    exprs_.push_back(expr);
+  }
+
+  void clearTraversalInfo();
+
+  std::vector<SegmentedGroup*> getNeighbors();
+
+  // TODO: May want to sort this based on size of connections between this and
+  // neighbors as well as if the connection is an output of the fusion (has to
+  // be saved to gmem anyways)
+  std::vector<std::pair<SegmentedGroup*, SegmentedEdge*>> getNeighborsPair();
```

Comment:
please replace std::pair uses with dedicated structs

Reply:
Updated to use `NeighborGroup` struct. Thanks!